_posts/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors.md

```diff
@@ -0,0 +1,155 @@
+          "name": "hnsw",
+          "engine": "faiss"
+        }
+      }
+    }
+  }
+}
+```
+
+Binary vectors in OpenSearch use Hamming distance for indexing and search.
+
+Next, pack the binary vector into a byte format for both indexing and searching. Other than that, using binary vectors is similar to working with FP32 vectors. In the following example, you'll index two documents with vector values of `[0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0]` and `[0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1]`:
+
+```json
+PUT _bulk
+{"index": {"_index": "test-binary-hnsw", "_id": "1"}}
+{"my_vector": [7, 8]}
+{"index": {"_index": "test-binary-hnsw", "_id": "2"}}
+{"my_vector": [10, 11]}
+```
+
+Finally, search for the vector closest to a query vector of `[0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0]`:
+
+```json
+GET /test-binary-hnsw/_search
+{
+  "size": 1,
+  "query": {
+    "knn": {
+      "my_vector": {
+        "vector": [108, -116],
+        "k": 1
+      }
+    }
+  }
+}
+```
+
+## Performance comparison
+
+Now, let's look at the resource savings when you're using binary vectors. In our benchmarking tests, we observed similar ingestion speeds and query times between FP32 and binary vectors, even while using 8x less powerful hardware for binary vectors, as shown in the following image.
+
+<img src="/assets/media/blog-images/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors/pic3.png" alt="performance" class="img-centered"/>
+
+### Cluster setup
+
+Our benchmark used a randomly generated 100-million vector dataset with 768 dimensions, comparing FP32 and binary vectors. The clusters were identical, except for data nodes: binary vectors used nodes that were 2x smaller and 4x fewer, leading to an 86% cost reduction. The following table outlines the benchmarking setup.
+
+<img src="/assets/media/blog-images/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors/table1.png" alt="performance" class="img-centered"/>
+
+### Performance results
+
+Even on 8x smaller hardware, binary vectors delivered comparable indexing speeds and query times to FP32 vectors on more powerful machines. With binary vectors, memory usage was reduced by 92%, and storage by 97%, producing significant savings. The results are presented in the following table.
+
+<img src="/assets/media/blog-images/2024-10-30-lower-your-cost-on-opensearch-using-binary-vectors/table2.png" alt="performance" class="img-centered"/>
+
+### Accuracy
+
+In terms of recall, you can expect around 0.97 recall compared to exact search. OpenSearch uses the HNSW algorithm for approximate nearest neighbor searches, but accuracy depends on your dataset. Some models produce binary vectors with high accuracy: for example, [Cohere Embed v3](https://cohere.com/blog/int8-binary-embeddings) reports a 94.7% match in search quality compared to FP32 embeddings. Thus, when using a model producing quality binary embeddings, binary vectors can reach nearly the same accuracy as FP32 vectors.
+
```

Comment:
Line 129: Should we define HNSW, or are we treating it as commonly understood (fine as is if the latter)?

Reply:
I think it's not necessary because this blog assumes familiarity with the algorithm.